keps/sig-network/2004-topology-aware-routing/README.md

```diff
@@ -0,0 +1,602 @@
+
+In this KEP we are going to focus on avoiding cross-zone traffic when in-zone
+endpoints would suffice. We're attempting to provide a simple and more
+automatic approach to topology aware routing. This API will still allow users
+to require traffic to stay within a zone, or simply to indicate that they
+prefer to keep traffic in the same zone if there's sufficient capacity. With
+this approach no new API fields are required, we can accomplish this simply
+with a new annotation on Service.
+
+
+### Goals
+- Provide a simple way for users to indicate their preference for keeping
+  traffic in zone.
+- Provide different approaches to users who have a strong desire over the
+  routing policy:
+  - Balanced: Use the original approach to ensure traffic is distributed across
+    endpoints for a Service. This may result in distribution across a subset of
+    endpoints.
+  - RequireZone: Only route to endpoints within the same zone.
+  - PreferZone: Prefer routing to endpoints within the same zone when they are
+    available and it makes sense. If there is an insufficient number of
+    endpoints in a zone, traffic may be routed to endpoints in other zones.
+- Use the standardized topology label `topology.kubernetes.io/zone` to derive
+  the zones of nodes and endpoints.
+- Use EndpointSlice subsetting as the primary mechanism for topology aware
+  routing.
+  - Improve kube-proxy scalability by limiting the number of endpoints it needs
+    to process.
+- Minimize churn of EndpointSlices while doing topology aware distribution.
+- Minimize the number of new EndpointSlices required.
+- Provide a simple API that requires minimal configuration for most users.
+
+### Non-Goals
+- Real-time distribution rebalancing based on traffic load or distribution
+  feedback or metrics.
+- Providing a way for kube-proxy to consume subsets of EndpointSlices. (This is
+  covered by [KEP 2030](https://github.com/kubernetes/enhancements/issues/2030)).
+- Providing a way to require Node local traffic ([this is a separate
+  KEP](https://github.com/kubernetes/enhancements/pull/1944)).
+- Multi-cluster topology aware routing (this same pattern may be useful there
+  though).
+- Region based topology aware routing (this may come later).
+- Ensuring that endpoints are distributed evenly across zones.
+
+## Proposal
+
+### Terminology
+
+- **EndpointSlice Group**: One or multiple EndpointSlices that will be consumed
+  by the same set of zones.
+- **Endpoints**: In this KEP, “endpoints” are used to describe individual Pod
+  IPs. These are not references to Kubernetes Endpoints resources.
+
+### Configuration
+
+This proposal builds off of the [EndpointSlice API](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/).
+
+We propose a new Service annotation: `service.kubernetes.io/same-zone`
+with three options:
+- `Balanced`: This represents the existing approach that is already used, no
```

Comment:
Comment got lost: I am afraid "balanced" implies more than we actually guarantee. Would "Random" be a better choice here? We don't guarantee to balance traffic, just spread it.

Reply:
What about "Unaware"?

Reply:
To go a slightly different direction here, let's say we had a different field/annotation here that encompassed both zone and region and supported values like `PreferZone`. 

Maybe that annotation is something like `endpointslice.kubernetes.io/slicing-policy` and maybe the supported values are:

- One of `None`,  `Random`, `Shared`, or `Balanced`
- `PreferZone`
- `RequireZone`

I don't think `same-zone: Random` or `same-zone: Balanced` make a lot of sense in comparison. This is a bit of a larger change though and changes the scope of this KEP + config from solving `topology aware routing` to defining a controller subsetting approach that helps implement topology aware routing.